### PR TITLE
fix(core): add lock for attachment

### PR DESCRIPTION
**Pull request by xujianhai666**

### CODE REVIEW

**Comment by imxyb** [comment]
why not use sync.Map?

**Comment by zouyx** [comment]
i have the same question.

File: protocol/invocation/rpcinvocation.go [hunk 1]
```diff
@@ -91,6 +93,8 @@ func (r *RPCInvocation) AttachmentsByKey(key string, defaultValue string) string
 }
 
 func (r *RPCInvocation) SetAttachments(key string, value string) {
+	r.lock.Lock()
```
Comment (AlexStocks) [inline protocol/invocation/rpcinvocation.go]: it seems that you just use the lock for writing. if u reading the attachment, u should use lock too.

Comment (xujianhai666) [reply protocol/invocation/rpcinvocation.go]: @AlexStocks okay.

File: protocol/invocation/rpcinvocation.go [hunk 2]
```diff
@@ -37,6 +38,7 @@ type RPCInvocation struct {
 	callBack       interface{}
 	attachments    map[string]string
 	invoker        protocol.Invoker
+	lock           sync.Mutex
```
Comment (fangyincheng) [inline protocol/invocation/rpcinvocation.go]: should it be a rwmutex?

Comment (hxmhlt) [reply protocol/invocation/rpcinvocation.go]: agree

Comment (xujianhai666) [reply protocol/invocation/rpcinvocation.go]: yes, done @fangyincheng

**Comment by xujianhai666** [comment]
> why not use sync.Map?

@imxyb @zouyx sync.map is too terrible for humable readable. and for official comment, if u can use map to deal with it, do not use sync.map

File: protocol/invocation/rpcinvocation.go [hunk 3]
```diff
@@ -80,6 +82,8 @@ func (r *RPCInvocation) Attachments() map[string]string {
 }
 
 func (r *RPCInvocation) AttachmentsByKey(key string, defaultValue string) string {
+	r.lock.Lock()
+	defer r.lock.Unlock()
```
Comment (fangyincheng) [inline protocol/invocation/rpcinvocation.go]: func (rw *RWMutex) RLock()

func (rw *RWMutex) RUnlock()

Comment (xujianhai666) [reply protocol/invocation/rpcinvocation.go]: @fangyincheng done

**Pull request merged**